### PR TITLE
Initial version of code for supporting API rate limiting - WIP

### LINKED ISSUE #249: Add Rate Limiting Support
We would like to add support for v3 Web API rate limits, as described [here](https://sendgrid.com/docs/API_Reference/Web_API_v3/How_To_Use_The_Web_API_v3/rate_limits.html).

**_How would you like this to be implemented?_** Or please give a thumbs up to express your interest in us implementing to move this task up our queue.

PRs are always welcome as well :)

### PR DESCRIPTION
**Pull request by waseem18**
This PR is for initial code review and is still  WIP. Will add more commits based on the feedback and answers from maintainers for few of my questions.

Fixes #249 

- Added variables `rate_limit_retry`, `rate_limit_sleep`, `RATE_LIMIT_RESPONSE_CODE` in the constructor of `SendGridAPIClient` class. 
- Added method `attempt` in the class `SendGridAPIClient`
- Added initial test file named `test_rate_limit.py` which calls `attempt` method.

@thinkingserious I had a couple of questions while implementing and put up those questions as single line comments in the code - I hope to get advise or answers to those questions 

Thanks

### CODE REVIEW

File: test/test_rate_limit.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    import unittest2 as unittest
+except ImportError:
+    import unittest
+
+
+class UnitTests(unittest.TestCase):
+
+    def test_rate_limit(self):
+        sg = sendgrid.SendGridAPIClient(apikey=os.environ.get('SENDGRID_API_KEY'))
+        request_obj = sg.client.access_settings.activity
```
Comment (waseem18) [inline test/test_rate_limit.py]: @thinkingserious  I checked `python-http-client` repository but couldn't find methods like `set_method` or `get_method`. So how would the user specify which method he wants to use while making the API call? Through parameter?

File: sendgrid/sendgrid.py [hunk 2]
```diff
@@ -73,6 +78,17 @@ def _get_default_headers(self):
     def reset_request_headers(self):
         self.client.request_headers = self._get_default_headers()
 
+    def attempt(self, request_obj):
+        for i in range(0, self.rate_limit_retry):
+            # For now assuming that the method is 'GET'
+            response = request_obj.get()
```
Comment (waseem18) [inline sendgrid/sendgrid.py]: Assumed (for the initial code review) that the method is `GET` as there's no method in `python-http-client` such as `set_method` or `get_method` as present in `java-http-client`.

Comment (thinkingserious) [reply sendgrid/sendgrid.py]: Here are the supported HTTP verbs: https://github.com/sendgrid/python-http-client/blob/master/python_http_client/client.py#L88

File: test/test_rate_limit.py (hunk 1 above)
Comment (thinkingserious) [reply test/test_rate_limit.py]: The `python-http-client` uses a [fluent interface](https://sendgrid.com/blog/using-python-to-implement-a-fluent-interface-to-any-rest-api/) with [these methods](https://github.com/sendgrid/python-http-client/blob/master/python_http_client/client.py#L88).

**Comment by waseem18** [comment]
@thinkingserious Added [new commit ](https://github.com/sendgrid/sendgrid-python/pull/551/commits/e6419532f3365beef2c7ce7f8aad4701e2f42f68) Can you please take a look at `make_request` method? If the flow is fine I'll wrap up the test cases.

Any feedback will be helpful.

Thanks

**Comment by thinkingserious** [comment]
Thanks @waseem18!

This PR is now on our backlog for a code review.

File: test/test_rate_limit.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        query_params = {}
+        data = {}
+        request_obj = sg.client.access_settings.activity
+        response = sg.make_request(request_obj,
+                              query_params=query_params,
+                              method="get",
+                              request_body=data)
+        print(response.status_code)
+        print(response.headers)
+        print(response.body)
```
Comment (thinkingserious) [inline test/test_rate_limit.py]: Unless I'm reading this wrong, I believe we need to add some assertions here to validate the test.

File: sendgrid/sendgrid.py [hunk 4]
```diff
@@ -73,6 +77,27 @@ def _get_default_headers(self):
...
+        :return: response of given request object
+        """
+        for i in range(0, self.rate_limit_retry):
+            response = getattr(request_obj, method)(request_body=request_body, query_params=query_params)
+            if response.status_code == self.RATE_LIMIT_RESPONSE_CODE:
+                rate_limit_reset = response.headers['X-Ratelimit-Reset']
+                time_to_sleep = int(rate_limit_reset) - int(time.time())
+                time.sleep(time_to_sleep)
+            else:
+                return response
```
Comment (thinkingserious) [inline sendgrid/sendgrid.py]: Should we throw an exception if the retry strategy fails?

File: sendgrid/sendgrid.py [hunk 5]
```diff
@@ -73,6 +77,27 @@ def _get_default_headers(self):
     def reset_request_headers(self):
         self.client.request_headers = self._get_default_headers()
 
+    def make_request(self, request_obj, query_params=None, method="get", request_body=None):
```
Comment (thinkingserious) [inline sendgrid/sendgrid.py]: I believe this code should be incorporated here: https://github.com/sendgrid/python-http-client/blob/master/python_http_client/client.py#L158 and be configured in the constructor here: https://github.com/sendgrid/sendgrid-python/blob/master/sendgrid/sendgrid.py#L35

**Comment by waseem18** [comment]
Thanks for the review @thinkingserious . I'll take some time this week to resolve code reviews.

**Comment by thinkingserious** [comment]
Awesome, thank you @waseem18!

**Comment by waseem18** [comment]
Your welcome @thinkingserious - I'll address the reviews and add-in more code soon.

**Comment by thinkingserious** [comment]
Hi @waseem18,

Just checking in to see if there is anything I can do to help move this forward. Thanks!

**Comment by waseem18** [comment]
@thinkingserious I think it would be better if we can capture the rate limiting functionality in `python-http-client` than adding it here.

**Comment by childish-sambino** [comment]
Closing due to inactivity. Please ping here if work is to continue.

**Pull request closed without merging**